Make the token fetching process more robust

There are cases when an access token fetched from the DB fails on decryption. It should be said, that it's rather unlikely a token is stored in a malformed state as straight after storing we verify the storing operation by retrieving the just stored token, decrypting it and validating.

A potential solution might be adding a checksum that is persisted along the token. Later, we could use it on retrieval to validate if everything went fine. If not, another fetching attempt can be made. In case the re-fetching did not change much an error should be raised.

Another route to check may be the shape of the token after the encryption and base64 encoding. Specifically, I'm wondering if it does not contain characters that are prone to problems from the moment we pass it to the update query to the moment it ends up in the DB.

**Additional info:**
As far as I remember the error below is not the only failure that we have observed so far. But this has just happened today on renkulab:
```
2023-05-04 08:04:53,217 ERROR application - Associating token with projectId: 7 failed
java.lang.RuntimeException: AccessToken decryption failed
	at io.renku.tokenrepository.repository.AccessTokenCryptoImpl$$anonfun$meaningfulError$lzycompute$1.applyOrElse(AccessTokenCrypto.scala:108)
	at io.renku.tokenrepository.repository.AccessTokenCryptoImpl$$anonfun$meaningfulError$lzycompute$1.applyOrElse(AccessTokenCrypto.scala:107)
	at cats.ApplicativeError.$anonfun$recoverWith$1(ApplicativeError.scala:172)
	at flatMap @ io.renku.tokenrepository.repository.AccessTokenCryptoImpl.decrypt(AccessTokenCrypto.scala:56)
	at recoverWith$extension @ io.renku.metrics.MetricsRegistry$.apply(MetricsRegistry.scala:45)
	at map @ io.renku.tokenrepository.repository.creation.TokensCreatorImpl.$anonfun$decrypt$1(TokensCreator.scala:76)
	at modify @ io.renku.metrics.MetricsRegistry$EnabledMetricsRegistry.register(MetricsRegistry.scala:70)
	at flatten$extension @ skunk.util.Pool$.$anonfun$ofF$2(Pool.scala:103)
	at modify @ io.renku.metrics.MetricsRegistry$EnabledMetricsRegistry.register(MetricsRegistry.scala:70)
	at flatten$extension @ skunk.util.Pool$.$anonfun$ofF$2(Pool.scala:103)
Caused by: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
	at java.base/com.sun.crypto.provider.CipherCore.unpad(Unknown Source)
	at java.base/com.sun.crypto.provider.CipherCore.fillOutputBuffer(Unknown Source)
	at java.base/com.sun.crypto.provider.CipherCore.doFinal(Unknown Source)
	at java.base/com.sun.crypto.provider.AESCipher.engineDoFinal(Unknown Source)
	at java.base/javax.crypto.Cipher.doFinal(Unknown Source)
	at io.renku.crypto.AesCrypto.$anonfun$decodeAndDecrypt$1(AesCrypto.scala:63)
	at cats.ApplicativeError.catchNonFatal(ApplicativeError.scala:269)
	at cats.ApplicativeError.catchNonFatal$(ApplicativeError.scala:268)
	... 7 common frames omitted
```

and

```
2023-05-04 08:04:53,231 ERROR application - Associating token with projectId: 3436 failed
java.lang.RuntimeException: AccessToken decryption failed
	at io.renku.tokenrepository.repository.AccessTokenCryptoImpl$$anonfun$meaningfulError$lzycompute$1.applyOrElse(AccessTokenCrypto.scala:108)
	at io.renku.tokenrepository.repository.AccessTokenCryptoImpl$$anonfun$meaningfulError$lzycompute$1.applyOrElse(AccessTokenCrypto.scala:107)
	at cats.ApplicativeError.$anonfun$recoverWith$1(ApplicativeError.scala:172)
	at flatMap @ io.renku.tokenrepository.repository.AccessTokenCryptoImpl.decrypt(AccessTokenCrypto.scala:56)
	at recoverWith$extension @ io.renku.metrics.MetricsRegistry$.apply(MetricsRegistry.scala:45)
	at map @ io.renku.tokenrepository.repository.creation.TokensCreatorImpl.$anonfun$decrypt$1(TokensCreator.scala:76)
	at modify @ io.renku.metrics.MetricsRegistry$EnabledMetricsRegistry.register(MetricsRegistry.scala:70)
	at flatten$extension @ skunk.util.Pool$.$anonfun$ofF$2(Pool.scala:103)
	at modify @ io.renku.metrics.MetricsRegistry$EnabledMetricsRegistry.register(MetricsRegistry.scala:70)
	at flatten$extension @ skunk.util.Pool$.$anonfun$ofF$2(Pool.scala:103)
Caused by: io.circe.ParsingFailure: expected json value got '��u��...' (line 1, column 1)
```

--- a/graph-commons/src/main/scala/io/renku/config/ConfigLoader.scala
+++ b/graph-commons/src/main/scala/io/renku/config/ConfigLoader.scala
@@ -21,9 +21,11 @@ package io.renku.config
 import cats.MonadThrow
 import cats.syntax.all._
 import com.typesafe.config.Config
+import io.renku.crypto.AesCrypto.Secret
 import io.renku.tinytypes._
 import pureconfig._
 import pureconfig.error.{CannotConvert, ConfigReaderFailures}
+import scodec.bits.ByteVector
 
 abstract class ConfigLoader[F[_]: MonadThrow] {
 
@@ -76,4 +78,17 @@ object ConfigLoader {
           }
           .getOrElse(Left(CannotConvert(stringValue, ttApply.getClass.toString, "Not an int value")))
       }
+
+  implicit def base64ByteVectorReader: ConfigReader[ByteVector] =
+    ConfigReader.fromString { str =>
+      ByteVector
+        .fromBase64Descriptive(str)
+        .left
+        .map(err => CannotConvert(str, "ByteVector", s"Cannot convert base64: $err"))
+    }
+
+  implicit def secretReader: ConfigReader[Secret] =
+    base64ByteVectorReader.emap(bv =>
+      Secret(bv).leftMap(err => CannotConvert(bv.toBase64, "Secret", s"Cannot create AES secret: $err"))
+    )
 }

--- a/graph-commons/src/main/scala/io/renku/crypto/AesCrypto.scala
+++ b/graph-commons/src/main/scala/io/renku/crypto/AesCrypto.scala
@@ -70,5 +70,11 @@ object AesCrypto {
       else Left(s"Expected 16 bytes, but got ${value.length}")
 
     def unsafe(value: ByteVector): Secret = apply(value).fold(sys.error, identity)
+
+    def fromBase64(b64: String): Either[String, Secret] =
+      ByteVector.fromBase64Descriptive(b64).flatMap(apply)
+
+    def unsafeFromBase64(b64: String): Secret =
+      fromBase64(b64).fold(sys.error, identity)
   }
 }

--- a/graph-commons/src/main/scala/io/renku/crypto/AesCrypto.scala
+++ b/graph-commons/src/main/scala/io/renku/crypto/AesCrypto.scala
@@ -34,13 +34,11 @@ abstract class AesCrypto[F[_]: MonadThrow, NONENCRYPTED, ENCRYPTED](
     secret: Secret
 ) {
 
-  private val base64Decoder    = Base64.getDecoder
-  private val base64Encoder    = Base64.getEncoder
-  private val algorithm        = "AES/CBC/PKCS5Padding"
-  private val key              = new SecretKeySpec(base64Decoder.decode(secret.value).takeWhile(_ != 10), "AES")
-  private val ivSpec           = new IvParameterSpec(new Array[Byte](16))
-  private val encryptingCipher = cipher(ENCRYPT_MODE)
-  private val decryptingCipher = cipher(DECRYPT_MODE)
+  private val base64Decoder = Base64.getDecoder
+  private val base64Encoder = Base64.getEncoder
+  private val algorithm     = "AES/CBC/PKCS5Padding"
+  private def key           = new SecretKeySpec(base64Decoder.decode(secret.value).takeWhile(_ != 10), "AES")
+  private val ivSpec        = new IvParameterSpec(new Array[Byte](16))
 
   def encrypt(nonEncrypted: NONENCRYPTED): F[ENCRYPTED]
   def decrypt(encrypted:    ENCRYPTED):    F[NONENCRYPTED]
@@ -53,14 +51,14 @@ abstract class AesCrypto[F[_]: MonadThrow, NONENCRYPTED, ENCRYPTED](
 
   protected def encryptAndEncode(toEncryptAndEncode: String): F[String] = MonadThrow[F].catchNonFatal {
     new String(
-      base64Encoder.encode(encryptingCipher.doFinal(toEncryptAndEncode.getBytes(UTF_8))),
+      base64Encoder.encode(cipher(ENCRYPT_MODE).doFinal(toEncryptAndEncode.getBytes(UTF_8))),
       UTF_8
     )
   }
 
   protected def decodeAndDecrypt(toDecodeAndDecrypt: String): F[String] = MonadThrow[F].catchNonFatal {
     new String(
-      decryptingCipher.doFinal(base64Decoder.decode(toDecodeAndDecrypt.getBytes(UTF_8))),
+      cipher(DECRYPT_MODE).doFinal(base64Decoder.decode(toDecodeAndDecrypt.getBytes(UTF_8))),
       UTF_8
     )
   }

--- a/graph-commons/src/main/scala/io/renku/crypto/AesCrypto.scala
+++ b/graph-commons/src/main/scala/io/renku/crypto/AesCrypto.scala
@@ -19,10 +19,8 @@
 package io.renku.crypto
 
 import cats.MonadThrow
-import eu.timepit.refined.W
-import eu.timepit.refined.api.Refined
-import eu.timepit.refined.collection.MinSize
 import io.renku.crypto.AesCrypto.Secret
+import scodec.bits.ByteVector
 
 import java.nio.charset.StandardCharsets.UTF_8
 import java.util.Base64
@@ -37,7 +35,6 @@ abstract class AesCrypto[F[_]: MonadThrow, NONENCRYPTED, ENCRYPTED](
   private val base64Decoder = Base64.getDecoder
   private val base64Encoder = Base64.getEncoder
   private val algorithm     = "AES/CBC/PKCS5Padding"
-  private def key           = new SecretKeySpec(base64Decoder.decode(secret.value).takeWhile(_ != 10), "AES")
   private val ivSpec        = new IvParameterSpec(new Array[Byte](16))
 
   def encrypt(nonEncrypted: NONENCRYPTED): F[ENCRYPTED]
@@ -45,15 +42,12 @@ abstract class AesCrypto[F[_]: MonadThrow, NONENCRYPTED, ENCRYPTED](
 
   private def cipher(mode: Int): Cipher = {
     val c = Cipher.getInstance(algorithm)
-    c.init(mode, key, ivSpec)
+    c.init(mode, new SecretKeySpec(secret.value.toArray, "AES"), ivSpec)
     c
   }
 
   protected def encryptAndEncode(toEncryptAndEncode: String): F[String] = MonadThrow[F].catchNonFatal {
-    new String(
-      base64Encoder.encode(cipher(ENCRYPT_MODE).doFinal(toEncryptAndEncode.getBytes(UTF_8))),
-      UTF_8
-    )
+    base64Encoder.encodeToString(cipher(ENCRYPT_MODE).doFinal(toEncryptAndEncode.getBytes(UTF_8)))
   }
 
   protected def decodeAndDecrypt(toDecodeAndDecrypt: String): F[String] = MonadThrow[F].catchNonFatal {
@@ -65,5 +59,16 @@ abstract class AesCrypto[F[_]: MonadThrow, NONENCRYPTED, ENCRYPTED](
 }
 
 object AesCrypto {
-  type Secret = String Refined MinSize[W.`16`.T]
+  final class Secret private (val value: ByteVector) extends AnyVal {
+    override def toString: String = "Secret(***)"
+
+    def toBase64 = value.toBase64
+  }
+  object Secret {
+    def apply(value: ByteVector): Either[String, Secret] =
+      if (value.length == 16) Right(new Secret(value))
+      else Left(s"Expected 16 bytes, but got ${value.length}")
+
+    def unsafe(value: ByteVector): Secret = apply(value).fold(sys.error, identity)
+  }
 }

--- a/graph-commons/src/test/scala/io/renku/generators/CommonGraphGenerators.scala
+++ b/graph-commons/src/test/scala/io/renku/generators/CommonGraphGenerators.scala
@@ -26,7 +26,7 @@ import io.renku.config.certificates.Certificate
 import io.renku.config.sentry.SentryConfig
 import io.renku.config.sentry.SentryConfig.{Dsn, Environment}
 import io.renku.control.{RateLimit, RateLimitUnit}
-import io.renku.crypto.AesCrypto
+import io.renku.crypto.AesCrypto.Secret
 import io.renku.generators.Generators.Implicits._
 import io.renku.generators.Generators._
 import io.renku.graph.http.server.security.Authorizer.AuthContext
@@ -49,20 +49,20 @@ import io.renku.triplesstore._
 import org.http4s.Status
 import org.http4s.Status._
 import org.scalacheck.{Arbitrary, Gen}
+import scodec.bits.Bases.Alphabets
+import scodec.bits.ByteVector
 
-import java.nio.charset.StandardCharsets.UTF_8
-import java.util.Base64
 import scala.language.implicitConversions
 import scala.util.Try
 
 object CommonGraphGenerators {
 
-  implicit val aesCryptoSecrets: Gen[AesCrypto.Secret] =
-    stringsOfLength(16)
-      .map(_.getBytes(UTF_8))
-      .map(Base64.getEncoder.encode)
-      .map(new String(_, UTF_8))
-      .map(Refined.unsafeApply)
+  implicit val aesCryptoSecrets: Gen[Secret] =
+    Gen
+      .listOfN(32, Gen.hexChar)
+      .map(_.mkString.toLowerCase)
+      .map(ByteVector.fromValidHex(_, Alphabets.HexLowercase))
+      .map(Secret.unsafe)
 
   implicit val personalAccessTokens: Gen[PersonalAccessToken] = for {
     length <- Gen.choose(5, 40)

--- a/token-repository/src/main/scala/io/renku/tokenrepository/repository/AccessTokenCrypto.scala
+++ b/token-repository/src/main/scala/io/renku/tokenrepository/repository/AccessTokenCrypto.scala
@@ -24,7 +24,6 @@ import cats.syntax.all._
 import com.typesafe.config.{Config, ConfigFactory}
 import eu.timepit.refined.W
 import eu.timepit.refined.api.{RefType, Refined}
-import eu.timepit.refined.pureconfig._
 import eu.timepit.refined.string.MatchesRegex
 import io.circe._
 import io.circe.parser._

--- a/token-repository/src/test/scala/io/renku/tokenrepository/repository/AccessTokenCryptoSpec.scala
+++ b/token-repository/src/test/scala/io/renku/tokenrepository/repository/AccessTokenCryptoSpec.scala
@@ -20,24 +20,25 @@ package io.renku.tokenrepository.repository
 
 import cats.syntax.all._
 import com.typesafe.config.ConfigFactory
-import eu.timepit.refined.api.RefType
+import io.renku.config.ConfigLoader.ConfigLoadingException
 import io.renku.crypto.AesCrypto.Secret
 import io.renku.generators.CommonGraphGenerators._
 import io.renku.generators.Generators.Implicits._
 import io.renku.generators.Generators._
 import io.renku.http.client.AccessToken
+import io.renku.testtools.IOSpec
 import io.renku.tokenrepository.repository.AccessTokenCrypto.EncryptedAccessToken
 import org.scalatest.matchers.should
 import org.scalatest.prop.TableDrivenPropertyChecks
 import org.scalatest.wordspec.AnyWordSpec
+import scodec.bits.ByteVector
 
 import java.nio.charset.StandardCharsets.UTF_8
-import java.security.InvalidKeyException
 import java.util.Base64
 import scala.jdk.CollectionConverters._
 import scala.util.{Failure, Success, Try}
 
-class AccessTokenCryptoSpec extends AnyWordSpec with should.Matchers with TableDrivenPropertyChecks {
+class AccessTokenCryptoSpec extends AnyWordSpec with should.Matchers with TableDrivenPropertyChecks with IOSpec {
 
   "encrypt/decrypt" should {
 
@@ -67,7 +68,7 @@ class AccessTokenCryptoSpec extends AnyWordSpec with should.Matchers with TableD
       val config = ConfigFactory.parseMap(
         Map(
           "projects-tokens" -> Map(
-            "secret" -> aesCryptoSecrets.generateOne.value
+            "secret" -> aesCryptoSecrets.generateOne.toBase64
           ).asJava
         ).asJava
       )
@@ -90,19 +91,15 @@ class AccessTokenCryptoSpec extends AnyWordSpec with should.Matchers with TableD
 
       val Failure(exception) = AccessTokenCrypto[Try](config)
 
-      exception            shouldBe a[InvalidKeyException]
-      exception.getMessage shouldBe "Invalid AES key length: 15 bytes"
+      exception          shouldBe a[ConfigLoadingException]
+      exception.getMessage should include("Expected 16 bytes, but got 15")
     }
   }
 
   private trait TestCase {
 
-    private val secret = new String(Base64.getEncoder.encode("1234567890123456".getBytes(UTF_8)), UTF_8)
-    val hookTokenCrypto = new AccessTokenCryptoImpl[Try](
-      RefType
-        .applyRef[Secret](secret)
-        .getOrElse(throw new IllegalArgumentException("Wrong secret"))
-    )
+    private val secret  = Secret.unsafe(ByteVector.view("1234567890123456".getBytes(UTF_8)))
+    val hookTokenCrypto = new AccessTokenCryptoImpl[Try](secret)
   }
 
   private lazy val tokenScenarios = Table(

--- a/token-repository/src/test/scala/io/renku/tokenrepository/repository/AccessTokenCryptoSpec.scala
+++ b/token-repository/src/test/scala/io/renku/tokenrepository/repository/AccessTokenCryptoSpec.scala
@@ -26,6 +26,7 @@ import io.renku.generators.CommonGraphGenerators._
 import io.renku.generators.Generators.Implicits._
 import io.renku.generators.Generators._
 import io.renku.http.client.AccessToken
+import io.renku.http.client.AccessToken.{PersonalAccessToken, UserOAuthAccessToken}
 import io.renku.testtools.IOSpec
 import io.renku.tokenrepository.repository.AccessTokenCrypto.EncryptedAccessToken
 import org.scalatest.matchers.should
@@ -59,6 +60,23 @@ class AccessTokenCryptoSpec extends AnyWordSpec with should.Matchers with TableD
 
       decryptException            shouldBe an[Exception]
       decryptException.getMessage shouldBe "AccessToken decryption failed"
+    }
+
+    "decrypt existing values" in {
+      val usedSecret  = Secret.unsafeFromBase64("YWJjZGVmZzEyMzQ1Njc4OQ==")
+      val tokenCrypto = new AccessTokenCryptoImpl[Try](usedSecret)
+      val token1      = PersonalAccessToken("lhcrr5dkerm69osrrujlubkkw0ysx8io0e")
+      val encToken1 = EncryptedAccessToken
+        .from("jmamuPCm4R0Rr/ZuVRrYffwBHWdJt8siwuVYJ7WVrLgDIR6RzcCJcpuuvCWcVnlPRsXi2wFHfM+OBOsbt2erZQ==")
+        .fold(throw _, identity)
+
+      val token2 = UserOAuthAccessToken("q2f4t4ph7atcfh08eau1g35")
+      val encToken2 = EncryptedAccessToken
+        .from("QXK4EOJLgdqQdh8MLkS+vCtsJFU0wsnpD9QZuu5qyWROFCzFfUr81xv6f1mN1r3T")
+        .fold(throw _, identity)
+
+      tokenCrypto.decrypt(encToken1) shouldBe Success(token1)
+      tokenCrypto.decrypt(encToken2) shouldBe Success(token2)
     }
   }
 

--- a/webhook-service/src/main/scala/io/renku/webhookservice/crypto/HookTokenCrypto.scala
+++ b/webhook-service/src/main/scala/io/renku/webhookservice/crypto/HookTokenCrypto.scala
@@ -23,7 +23,6 @@ import cats.syntax.all._
 import com.typesafe.config.{Config, ConfigFactory}
 import eu.timepit.refined.W
 import eu.timepit.refined.api.{RefType, Refined}
-import eu.timepit.refined.pureconfig._
 import eu.timepit.refined.string.MatchesRegex
 import io.circe.parser._
 import io.circe.{Decoder, HCursor, Json}

--- a/webhook-service/src/test/scala/io/renku/webhookservice/crypto/HookTokenCryptoSpec.scala
+++ b/webhook-service/src/test/scala/io/renku/webhookservice/crypto/HookTokenCryptoSpec.scala
@@ -58,6 +58,18 @@ class HookTokenCryptoSpec extends AnyWordSpec with should.Matchers {
       decryptException            shouldBe an[Exception]
       decryptException.getMessage shouldBe "HookToken decryption failed"
     }
+
+    "decrypt existing values" in {
+      val usedSecret      = Secret.unsafeFromBase64("YWJjZGVmZzEyMzQ1Njc4OQ==")
+      val hookTokenCrypto = new HookTokenCryptoImpl[Try](usedSecret)
+
+      val plain =
+        hookTokenCrypto.decrypt(
+          SerializedHookToken.from("4TV8A81+1+U3dfa8sVO9TUuMvGHLaerySYpxEIkVT/U=").fold(throw _, identity)
+        )
+
+      plain shouldBe Success(HookToken(123456))
+    }
   }
 
   "apply" should {

--- a/webhook-service/src/test/scala/io/renku/webhookservice/crypto/HookTokenCryptoSpec.scala
+++ b/webhook-service/src/test/scala/io/renku/webhookservice/crypto/HookTokenCryptoSpec.scala
@@ -19,7 +19,7 @@
 package io.renku.webhookservice.crypto
 
 import com.typesafe.config.ConfigFactory
-import eu.timepit.refined.api.RefType
+import io.renku.config.ConfigLoader.ConfigLoadingException
 import io.renku.crypto.AesCrypto.Secret
 import io.renku.generators.CommonGraphGenerators._
 import io.renku.generators.Generators.Implicits._
@@ -29,9 +29,9 @@ import io.renku.webhookservice.crypto.HookTokenCrypto.SerializedHookToken
 import io.renku.webhookservice.model.HookToken
 import org.scalatest.matchers.should
 import org.scalatest.wordspec.AnyWordSpec
+import scodec.bits.ByteVector
 
 import java.nio.charset.StandardCharsets.UTF_8
-import java.security.InvalidKeyException
 import java.util.Base64
 import scala.jdk.CollectionConverters._
 import scala.util.{Failure, Success, Try}
@@ -67,7 +67,7 @@ class HookTokenCryptoSpec extends AnyWordSpec with should.Matchers {
         Map(
           "services" -> Map(
             "gitlab" -> Map(
-              "hook-token-secret" -> aesCryptoSecrets.generateOne.value
+              "hook-token-secret" -> aesCryptoSecrets.generateOne.toBase64
             ).asJava
           ).asJava
         ).asJava
@@ -93,18 +93,14 @@ class HookTokenCryptoSpec extends AnyWordSpec with should.Matchers {
 
       val Failure(exception) = HookTokenCrypto[Try](config)
 
-      exception            shouldBe a[InvalidKeyException]
-      exception.getMessage shouldBe "Invalid AES key length: 15 bytes"
+      exception          shouldBe a[ConfigLoadingException]
+      exception.getMessage should include("Expected 16 bytes, but got 15")
     }
   }
 
   private trait TestCase {
 
-    private val secret = new String(Base64.getEncoder.encode("1234567890123456".getBytes("utf-8")), "utf-8")
-    val hookTokenCrypto = new HookTokenCryptoImpl[Try](
-      RefType
-        .applyRef[Secret](secret)
-        .getOrElse(throw new IllegalArgumentException("Wrong secret"))
-    )
+    private val secret  = Secret.unsafe(ByteVector.fromValidHex(List.fill(4)("caffee00").mkString))
+    val hookTokenCrypto = new HookTokenCryptoImpl[Try](secret)
   }
 }

